packages/loader/driver-utils/src/mapWithExpiration.ts

```diff
@@ -0,0 +1,115 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+
+/**
+ * An extension of Map that expires (deletes) entries after a period of inactivity.
+ * The policy is based on the last time a key was written to.
+ */
+export class MapWithExpiration<TKey, TValue> extends Map<TKey, TValue> {
+    /** Timestamps (as epoch ms numbers) of when each key was last refreshed */
+    private readonly lastRefreshedTimes = new Map<TKey, number>();
+
+    constructor(
+        private readonly expiryMs: number,
+    ) {
+        super();
+    }
+
+    private refresh(key: TKey): void {
+        this.lastRefreshedTimes.set(key, (new Date()).valueOf());
+    }
+
+    /**
+     * Returns true if the key is present and expired, false if it's not expired, and undefined if it's not found
+     * If cleanUp is passed as true, then delete any expired entry before returning.
+     */
+    private checkExpiry(key: TKey, cleanUp: boolean = false): boolean | undefined {
+        const refreshTime = this.lastRefreshedTimes.get(key);
+        assert((refreshTime !== undefined) === super.has(key), "freshness map out of sync");
+
+        if (refreshTime === undefined) {
+            return undefined;
+        }
+        const expired = (new Date()).valueOf() - refreshTime >= this.expiryMs;
+        if (expired && cleanUp) {
+            this.delete(key);
+        }
+        return expired;
+    }
+
+    private clearExpiredEntries() {
+        // forEach clears out any expired entries
+        this.forEach(() => {});
+    }
+
+    get size(): number {
+        this.clearExpiredEntries();
+        return super.size;
+    }
+
+    has(key: TKey): boolean {
+        this.checkExpiry(key, true /* cleanUp */);
+        return super.has(key);
+    }
+
+    get(key: TKey): TValue | undefined {
+        this.checkExpiry(key, true /* cleanUp */);
+        return super.get(key);
+    }
+
+    set(key: TKey, value: TValue): this {
+        // Sliding window expiration policy (on write)
+        this.refresh(key);
+        return super.set(key, value);
+    }
+
+    delete(key: TKey): boolean {
+        this.lastRefreshedTimes.delete(key);
+        return super.delete(key);
+    }
+
+    clear(): void {
+        this.lastRefreshedTimes.clear();
+        super.clear();
+    }
+
+    forEach(callbackfn: (value: TValue, key: TKey, map: Map<TKey, TValue>) => void, thisArg?: any): void {
```

Comment:
I don't really understand how `thisArg` works, so want to write a test to cover that once I learn how to even use it :P